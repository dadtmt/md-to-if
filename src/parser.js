import SimpleMarkdown from 'simple-markdown'

// String -> [String]
export const matchBraces = source => /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = source => /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = source => /^\|([\s\S]*?)\]/.exec(source)

const command = {
  order: 0,
  match: matchBraces,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

// String -> [Content]
const parser = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent,
})

export default parser
