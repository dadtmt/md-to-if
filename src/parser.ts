import SimpleMarkdown, { SingleASTNode } from 'simple-markdown'

// String -> [String]
export const matchBraces = (source: string): RegExpExecArray | null =>
  /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = (source: string): RegExpExecArray | null =>
  /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = (source: string): RegExpExecArray | null =>
  /^\|([\s\S]*?)\]/.exec(source)

const parse: (capture: any, parse: any, state: any) => any = (
  capture,
  parse,
  state
) => ({ content: parse(capture[1], state) })

const command = {
  order: 0,
  match: matchBraces,
  parse
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse
}

// String -> [Content]
const parser: (source: string) => SingleASTNode[] = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent
})

export default parser
