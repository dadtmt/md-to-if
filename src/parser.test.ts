import './test/types.d'
import parser, { matchBracketPipe, matchPipeBracket } from './parser'
import { ASTNode } from 'simple-markdown'

const { adventureGlobals } = global
const { adventureMd } = adventureGlobals

describe('matchBracketPipe', () => {
  it('match bracket and a pipe', () => {
    const source = '[ condition true | condition false ] }'

    expect(matchBracketPipe(source)).toMatchSnapshot()
  })
})

describe('matchPipeBracket', () => {
  it('match bracket and a pipe', () => {
    const source = '| condition false ] }'

    expect(matchPipeBracket(source)).toMatchSnapshot()
  })
})

describe('parser', () => {
  it('parses {} into command node', () => {
    const someCommandMd = `some content { some command } other content`
    const expected: ASTNode = {
      type: 'command',
      content: [
        {
          type: 'text',
          content: ' some command '
        }
      ]
    }
    const [mainContent] = parser(someCommandMd)
    const { content } = mainContent
    const secondNode = content[1]
    expect(secondNode).toEqual(expected)
  })
  it('parse markdown', () => {
    expect(parser(adventureMd)).toMatchSnapshot()
  })
})
