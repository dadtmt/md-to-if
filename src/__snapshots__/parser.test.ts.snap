// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matchBracketPipe match bracket and a pipe 1`] = `
Array [
  "[ condition true |",
  " condition true ",
]
`;

exports[`matchPipeBracket match bracket and a pipe 1`] = `
Array [
  "| condition false ]",
  " condition false ",
]
`;

exports[`parser parse markdown 1`] = `
Array [
  Object {
    "content": Array [
      Object {
        "content": "Choose your story",
        "type": "text",
      },
    ],
    "level": 1,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "This is a test adventure",
        "type": "text",
      },
      Object {
        "content": ", welcome",
        "type": "text",
      },
      Object {
        "content": ".",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "The space ship",
        "type": "text",
      },
    ],
    "level": 2,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "You are the brave pilot of this ship",
        "type": "text",
      },
      Object {
        "content": ".
The trip is long and boring",
        "type": "text",
      },
      Object {
        "content": ".",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Are you going to ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": "get food",
            "type": "text",
          },
        ],
        "target": "/cantina",
        "title": undefined,
        "type": "link",
      },
      Object {
        "content": " or to ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": "get a rest",
            "type": "text",
          },
        ],
        "target": "/bedroom",
        "title": undefined,
        "type": "link",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Cantina",
        "type": "text",
      },
    ],
    "level": 2,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": " show playedCount ",
        "type": "text",
      },
    ],
    "type": "command",
  },
  Object {
    "content": Array [
      Object {
        "content": " time you come here",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "The cantina is full with your crew having a good time ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " test playedCount equals val 1 ",
            "type": "text",
          },
        ],
        "type": "command",
      },
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": "trueCase",
            "type": "text",
          },
        ],
        "type": "trueCaseContent",
      },
      Object {
        "content": Array [
          Object {
            "content": "falseCase",
            "type": "text",
          },
        ],
        "type": "falseCaseContent",
      },
      Object {
        "content": " ",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "They are singing ",
        "type": "text",
      },
      Object {
        "content": "'oh my captain",
        "type": "text",
      },
      Object {
        "content": "' for the ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " show playedCount ",
            "type": "text",
          },
        ],
        "type": "command",
      },
      Object {
        "content": " times",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "You pick up a sandwich an put it in your bag ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " set bag sandwich val 1 ",
            "type": "text",
          },
        ],
        "type": "command",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "You find some pickles ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " set bag pickles val roll d10 ",
            "type": "text",
          },
        ],
        "type": "command",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "You have ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " show bag sandwich ",
            "type": "text",
          },
        ],
        "type": "command",
      },
      Object {
        "content": " with you",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "You have ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " show bag pickles ",
            "type": "text",
          },
        ],
        "type": "command",
      },
      Object {
        "content": " with you",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": " test playedCount equals val 1 ",
        "type": "text",
      },
    ],
    "type": "command",
  },
  Object {
    "content": Array [
      Object {
        "content": " ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " This is your first time here
    Will you come back ",
            "type": "text",
          },
          Object {
            "content": "? ",
            "type": "text",
          },
        ],
        "type": "trueCaseContent",
      },
      Object {
        "content": Array [
          Object {
            "content": " Should not you go somewhere else ",
            "type": "text",
          },
          Object {
            "content": "?
or what ",
            "type": "text",
          },
          Object {
            "content": "? ",
            "type": "text",
          },
        ],
        "type": "falseCaseContent",
      },
      Object {
        "content": " ",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Drink",
        "type": "text",
      },
    ],
    "level": 3,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "You have a whisky or a milkhake",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Whisky",
        "type": "text",
      },
    ],
    "level": 4,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "You feel dipsy",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Milkshake",
        "type": "text",
      },
    ],
    "level": 4,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "Delicious strawberry one",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Eat",
        "type": "text",
      },
    ],
    "level": 3,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "Some eggs",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "Bedroom",
        "type": "text",
      },
    ],
    "level": 2,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": "Crazy major droid shoots you ",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": " describe droid ",
        "type": "text",
      },
      Object {
        "align": Array [
          null,
          null,
          null,
          null,
        ],
        "cells": Array [
          Array [
            Array [
              Object {
                "content": "roll D6",
                "type": "text",
              },
            ],
            Array [
              Object {
                "content": "45",
                "type": "text",
              },
            ],
            Array [
              Object {
                "content": "24",
                "type": "text",
              },
            ],
            Array [
              Object {
                "content": "41",
                "type": "text",
              },
            ],
          ],
        ],
        "header": Array [
          Array [
            Object {
              "content": "CC",
              "type": "text",
            },
          ],
          Array [
            Object {
              "content": "CT",
              "type": "text",
            },
          ],
          Array [
            Object {
              "content": "F",
              "type": "text",
            },
          ],
          Array [
            Object {
              "content": "E",
              "type": "text",
            },
          ],
        ],
        "type": "table",
      },
    ],
    "type": "command",
  },
  Object {
    "type": "newline",
  },
  Object {
    "content": Array [
      Object {
        "content": "You roll a dice it gives ",
        "type": "text",
      },
      Object {
        "content": Array [
          Object {
            "content": " show roll d10 ",
            "type": "text",
          },
        ],
        "type": "command",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "The droid try to shoot you",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  Object {
    "content": Array [
      Object {
        "content": "The Droid shoots",
        "type": "text",
      },
    ],
    "level": 3,
    "type": "heading",
  },
  Object {
    "content": Array [
      Object {
        "content": " test roll d100 lte val droid CT ",
        "type": "text",
      },
    ],
    "type": "command",
  },
  Object {
    "content": Array [
      Object {
        "content": " you are shot ",
        "type": "text",
      },
    ],
    "type": "trueCaseContent",
  },
  Object {
    "content": Array [
      Object {
        "content": " You luckyly escape ",
        "type": "text",
      },
    ],
    "type": "falseCaseContent",
  },
  Object {
    "type": "newline",
  },
]
`;
